crates/polars-compute/src/rolling/no_nulls/mod.rs

```diff
@@ -93,8 +94,32 @@ where
         .map(|idx| {
             let (start, end) = det_offsets_fn(idx, window_size, len);
             let vals = unsafe { values.get_unchecked(start..end) };
+            let win_len = end - start;
+            let weights_slice = if start == 0 {
+                // Truncated at the start: take the last win_len weights
+                &weights[weights.len() - win_len..]
+            } else if end == len {
+                // Truncated at the end: take the first win_len weights
+                &weights[..win_len]
+            } else {
+                // Full window
+                weights
+            };
 
-            aggregator(vals, weights)
+            if normalize && win_len != window_size {
+                // Renormalize weights so they sum to 1
+                let wsum = weights_slice
+                    .iter()
+                    .copied()
+                    .fold(T::zero(), |acc, x| acc + x);
+                if wsum == T::zero() {
+                    panic!("Weighted mean is undefined if weights sum to 0");
+                }
+                let normed_weights: Vec<T> = weights_slice.iter().map(|&w| w / wsum).collect();
+                aggregator(vals, &normed_weights)
```

Comment:
I'm *really* not a fan of normalizing the weights for each call. Instead the inner aggregator if necessary (`mean` / `var`) should not assume the weights have been pre-normalized and calculate the total weight themselves.